### ISSUE TITLE
Add ability to really ignore some files

### ISSUE THREAD

**Issue by glassez** [arvidn/libtorrent#7143]
Currently, all `libtorrent` provides is only the "download priority", which literally affects only the download. `libtorrent` behaves greedily with respect to existing files, i.e. it handles already existing files in any case (both those that already existed before the torrent was added, and those previously downloaded by the torrent itself), even if they are set to "don't download" priority. But users expect different behavior. They need the ability to completely exclude some files from `libtorrent` handling (in fact, this is what they expect when they set "don't download" priority on existing files).
So it would be highly desirable to be able to set some "ignored" flag on a specific file that would allow you to completely ignore this file by doing the following:
1. When adding/checking a torrent, "ignored" files are completely skipped by `libtorrent`.
2. `libtorrent` doesn't even try to do anything with such files, neither download nor seed, their progress is obviously zero.
3. When moving torrent "ignored" files are also skipped from being moved.
4. When some currently handled file is switched to an "ignored" state, all work associated with it are terminated, it is closed, its progress is reset to zero, and possible data of cross-files pieces is moved to the parts file.

**Comment by glassez** [comment]
@arvidn, ping, in case you didn't notice it.

**Comment by glassez** [comment]
> A hack to get this behaviour (I believe) would be to rename the files to a random name, that don't exist, before adding the torrent.

It doesn't cover 4. and 5. above.


> It's not great though, and having some way to exclude files would be simpler and more robust.

Exactly!

**Comment by sserdda-liamE** [comment]
What use case would have you not want to seed a portion that you have?

I certainly don't expect "don't download" to mean, "Don't seed."  It just stops the client from "correcting" a file I don't want corrected.
"Ignore file completely" should be a separate option.

**Comment by glassez** [comment]
> What use case would have you not want to seed a portion that you have?
> I certainly don't expect "don't download" to mean, "Don't seed." It just stops the client from "correcting" a file I don't want corrected. "Ignore file completely" should be a separate option.

OK. I agree. (Honestly, I initially doubted about my suggestion above to replace `dont_download` with `ignore`. This should definitely be a separate option.)

**Comment by glassez** [comment]
> > 3. When moving torrent "ignored" files are also skipped from being moved.
> 
> This seems like arbitrary behavior.

This is just the same adequate behavior characterized by the term "ignore".


> If you have thousands of torrents running you're likely to forget which ones you had set up to ignore certain files, and this would leave you a mess of orphaned files on a mass move.

Sorry, it's a bad idea to put sticks in the wheels, preventing the addition of some feature, due to the fact that some user may use it incorrectly. By setting files as "ignored", you should be aware of the simplicity of this action and its consequences, so you should have a good reason to do it (e.g. you want to delete previously downloaded files).
At the very least, there may be some warning in case you are going to move a torrent that contains "ignored" files.


> This would be like if setting a file to Read Only left the original in place when you moved the directory, and setting it to Hidden would stop it from copying.

Sorry, I don't see any analogy with ignoring.
A correct analogy would be to apply Ctrl+A, Ctrl+C to the contents of a folder containing some hidden files. In this case, what you are suggesting would be equal to copying all files, including hidden ones, which is incorrect.


> Moving the torrent should probably move everything, including togglable "ignore" if it was ever in an "ignore = off" state.

I don't agree either. "Ignoring" should behave the same, depending only on its current state, without taking into account any background. Otherwise it doesn't make sense.

**Comment by sserdda-liamE** [comment]
> Sorry, it's a bad idea to put sticks in the wheels, preventing the addition of some feature, due to the fact that some user may use it incorrectly. By setting files as "ignored", you should be aware of the simplicity of this action and its consequences, so you should have a good reason to do it (e.g. you want to delete previously downloaded files). At the very least, there may be some warning in case you are going to move a torrent that contains "ignored" files.

You need that warning because there's zero discoverability of which way it would fall from just putting a file into an ignore state.  I would personally assume that the "directory" level action of moving an entire torrent would override any file-level settings.

You can't assume that a user starts by reading the man pages in their entirety, grasps the entire length, breadth, and depth of the program, and then loads up the program knowing every single aspect of every option.  Because the average user just loads it up, pokes at it until it breaks, and only then looks up what that option was supposed to do.

There is a good reason not to load a piece of software down with little-used options.  We don't do well with huge sets.  After a certain point it's better to break things off into separate utilities that can be learned individually as needed than sift through a tome of options that you're never going to use trying to winnow out what you mainly need to know.

Less chance of bugs, too.

What happens if you set a file to Ignore and there isn't enough disk space to make the .part files?  What do you do with the .part files when you set it to unignore?  Do you delete them?  There could be a second libtorrent instance also set to ignore that needs them.
Do you recheck on unignore?  What do you do if it doesn't recheck?  Would you really want to stop the entire torrent which had been happily partially seeding because the unignored file didn't recheck?  So you'd probably want a special error state where it puts the file back in ignore and continues.
What happens if the parts of the file that don't recheck are at the beginning and end of the file, which are covered by the .part files?  Do you correct the file from the .part files so you can seed?  That seems like it would be a no-no.  But you can seed the entire thing from the combination of .part files and the rest of the file... as long as you don't delete the .part files on setting unignore.

**Comment by glassez** [comment]
> You can't assume that a user starts by reading the man pages in their entirety, grasps the entire length, breadth, and depth of the program, and then loads up the program knowing every single aspect of every option. Because the average user just loads it up, pokes at it until it breaks, and only then looks up what that option was supposed to do.

It looks like it's you who complicates everything. Damn, this thing is "as plain as the nose on one's face".
It seems that you just don't like it for some subjective reasons (or you just don't understand its essence), and instead of just ignoring it (since it does not imply in any way influencing those who are not interested in it), you prefer to prevent it by all means.


> Less chance of bugs, too.

To implement nothing is "less chance of bugs" either.

Putting aside your complications, it's really simple from the perspective of its interface. It looks so natural to me that I couldn't even think that it needed to be explained to someone. The behavior of a torrent with some ignored files can be perceived as if the torrent does not contain these files at all.
Most of your "what..." questions above relate to the details of its implementation. If you reject the very idea of this feature, then there is no point in discussing them at all.

**Comment by sserdda-liamE** [comment]
> It looks like it's you who complicates everything. Damn, this thing is "as plain as the nose on one's face". It seems that you just don't like it for some subjective reasons (or you just don't understand its essence), and instead of just ignoring it (since it does not imply in any way influencing those who are not interested in it), you prefer to prevent it by all means.

I do not care one way or the other.  However, you have not yet come up with a single use case for this feature.  
Should libtorrent also have the ability to format a disk?  Should it also have an antivirus?  A text editor?  An init system?  A tower defense game?  An image viewer?  
Just because one can posit a feature does not mean it is good use of developer time and worth the maintenance overhead.

Google Chrome religiously removes little-used features for a reason.  It keeps the UI and codebase cleaner.
You are here asking for a feature that I have never needed.
If one has some need to hide from a torrent client the fact that one has a file, moving it in a File Explorer will do the trick.

 
> Putting aside your complications, it's really simple from the perspective of its interface. It looks so natural to me that I couldn't even think that it needed to be explained to someone. The behavior of a torrent with some ignored files can be perceived as if the torrent does not contain these files at all. 

You perceive that only because you have imagined the implementation you want first.  There is nothing natural about your interpretation, though.
If I set an email message to ignore and then delete my entire account, I do not expect the email message to pop out in public as the privacy features of the account vanish around it.  Ignoring within a structure does not imply ignoring when acting on the structure in its entirety.
Your desired implementation goes against norms, so it would not be normal for a person to assume it works that way.

**Comment by sserdda-liamE** [comment]
> I don't need to invent any use case. This issue is based on real requests from real users of a real application.

A feature which they have no use for is equally served by not implementing it.

**Comment by sledgehammer999** [comment]
If I understand correctly: The new `ignored` state can be thought of as "unmanaged files".
Possible use case?
A. Users just throw their torrents into the same folder which may result in conflicting files. They don't want to invoke hash checking because the files are really different and only share the filename. A hash-check will setup libtorrent to do operations on that file (eg truncate it) which will destroy the data present in it.
B. Same concept as A but the conflicting file comes from a later torrent that the user chooses to download.
In all cases, if the files marked as "don't download" have their torrents moved then libtorrent will move the marked file because it actually exists on disk. But this will result in errors in the torrent where the file isn't marked as "dont download" and not moved too. Also I wonder what will happen if multiple torrents share the same filename but only one of them is set to download. Will the others try to hash-check it and seed it if they see it as present on disk? If yes, this can easily turn into a lot of incosistencies in libtorrent and probably in errors or data deletion.

**Comment by anonghuser** [comment]
Hey, I'm not sure if there are notifications for issue being mentioned, so I'll ping.
The above mention gives a use-case that seems was being sought for this issue. 

tl;dr: I want to be able to delete a file from a torrent without having to force-recheck or redownload its edge pieces shared with other files.

Most commonly I need this for two reasons: 
- unwanted nfo/txt/whatever I accidentally left selected initially
- due to storage space limitations I need to download just a part of big torrents, then later delete some or all of it while setting other parts to download, etc.

**Comment by luzpaz** [comment]
bumping to refresh

**Comment by Phantom1472** [comment]
Hey, would really love this feature!

**Comment by asakura42** [comment]
Bump. Must have ability for large torrent files with many files inside it. It's not convenient to delete files manually and fall into errors (i use qbt tho).

**Comment by tooruu** [comment]
Please implement this!

**Comment by glassez** [comment]
> Thank you for your contributions.

You are welcome... 👎

**Comment by arvidn** [comment]
I wonder if it would be possible to re-define the meaning of `dont_download` to mean `ignore`

**Comment by arvidn** [comment]
one challenge is marking a file ignored after having downloaded at least one piece to it, and advertized having that piece to a peer. If we drop the piece (and the whole file) I can see 4 options:

1. if the peer supports `lt_dont_have`, advertize that we lost the piece
2. Disconnect all peers to force a bitfield refresh
3. Send an unsolicited bitfield message where the piece(s) have been removed and hope for the best
4. Don't tell the peer that we lost the piece and hope it won't request it. Once it does, fall back to 1-3 and reject the request

**Comment by glassez** [comment]
> one challenge is marking a file ignored after having downloaded at least one piece to it, and advertized having that piece to a peer.

I think we should assume that users are doing this anyway and will continue to do so (deleting previously downloaded files) when they need to for some reason (saving disk space etc.). We should just regard this as an unavoidable use case and improve its handling and UX. And since the other side may in any case encounter a situation where some piece that was previously reported available may end up being lost, all that remains for us is to handle it gracefully, if possible. As for the specific way to do this, I don't have a strict preference, as I'm not competent enough in these details.
